### PR TITLE
Implement first function for latitude weighting

### LINKED ISSUE #466: Latitude-weighted loss function
### Is your feature request related to a problem? Please describe.

_No response_

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by Jubeku**
## Description

Adding optional latitude weighting functions to loss calculations. 

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #466 

## Code Compatibility

-   [ ] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/weathergen/train/loss_calculator.py [hunk 1]
```diff
@@ -140,7 +154,8 @@ def _loss_per_loss_function(
 
         ctr_substeps = 0
         for mask_t in substep_masks:
-            assert mask_t.sum() == len(weights_locations) if weights_locations else True
+            if weights_locations is not None:
```
Comment (clessig) [inline src/weathergen/train/loss_calculator.py]: This is equivalent to the old version, not?

File: src/weathergen/train/loss_calculator.py [hunk 2]
```diff
@@ -105,6 +105,26 @@ def _get_weights(self, stream_info):
 
         return stream_info_loss_weight, weights_channels
 
+    def cosine_latitude(self, stream_data, forecast_offset, fstep, min_value=1e-3, max_value=1.0):
```
Comment (clessig) [inline src/weathergen/train/loss_calculator.py]: This function should be outside of the class

File: src/weathergen/train/loss_calculator.py [hunk 3]
```diff
@@ -105,6 +105,26 @@ def _get_weights(self, stream_info):
...
+        latitudes_radian = (
+            stream_data.target_coords_raw[forecast_offset + fstep][:, 0] * np.pi / 180
+        )
+        return (max_value - min_value) * np.cos(latitudes_radian) + min_value
+
+    def _get_location_weights(self, stream_info, stream_data, forecast_offset, fstep):
+        stream_info_location_weight = stream_info.get("location_weight", None)
```
Comment (clessig) [inline src/weathergen/train/loss_calculator.py]: stream_info_location_weight -> location_weight_name or location_weight_type

File: src/weathergen/train/loss_calculator.py [hunk 4]
```diff
@@ -105,6 +105,26 @@ def _get_weights(self, stream_info):
...
+        weights_locations_fct = (
+            getattr(self, stream_info_location_weight) if stream_info_location_weight else None
+        )
+        weights_locations = (
+            weights_locations_fct(stream_data, forecast_offset, fstep)
+            if weights_locations_fct
+            else None
+        )
+        weights_locations = weights_locations.to(device=self.device, non_blocking=True)
```
Comment (clessig) [inline src/weathergen/train/loss_calculator.py:114]: This will fail when weights_locations is None, i.e. when weights_locations_fct is None

File: src/weathergen/train/loss_calculator.py [hunk 5]
```diff
@@ -105,6 +105,26 @@ def _get_weights(self, stream_info):
...
 
+    def cosine_latitude(self, stream_data, forecast_offset, fstep, min_value=1e-3, max_value=1.0):
+        latitudes_radian = (
+            stream_data.target_coords_raw[forecast_offset + fstep][:, 0] * np.pi / 180
+        )
+        return (max_value - min_value) * np.cos(latitudes_radian) + min_value
+
+    def _get_location_weights(self, stream_info, stream_data, forecast_offset, fstep):
+        stream_info_location_weight = stream_info.get("location_weight", None)
+        weights_locations_fct = (
```
Comment (clessig) [inline src/weathergen/train/loss_calculator.py]: I would early out before this function.

```
if weights_locations_fct is None:
  return None
```

File: src/weathergen/train/loss.py [hunk 6]
```diff
@@ -135,9 +135,10 @@ def mse_channel_location_weighted(
     mask_nan = ~torch.isnan(target)
     pred = pred[0] if pred.shape[0] == 0 else pred.mean(0)
 
-    diff2 = torch.square(torch.where(mask_nan, target, 0) - torch.where(mask_nan, pred, 0)).mean(0)
-    wl = weights_points
-    loss_chs = ((diff2.transpose(1, 0) * wl).transpose(1, 0) if wl else diff2).mean(0)
+    diff2 = torch.square(torch.where(mask_nan, target, 0) - torch.where(mask_nan, pred, 0))
```
Comment (clessig) [inline src/weathergen/train/loss.py:138]: Thanks for the fix here. This was actually incorrect (and only worked since we didn't have weights).

**Review by clessig — COMMENTED** [review]
Looks overall good, just some minor comments. We should get this merged quite quickly and have separate PRs for adding more complex lat- and variable weighting schemes.

File: src/weathergen/train/loss_calculator.py (hunk 1 above)
Comment (Jubeku) [reply src/weathergen/train/loss_calculator.py]: For me the in-line if condition doesn't work if weights_locations is a tensor but checking if it's not None does. Does this make sense?

Comment (clessig) [reply src/weathergen/train/loss_calculator.py]: But would this work:

```
assert mask_t.sum() == len(weights_locations) if weights_locations is not None else True
```

Comment (Jubeku) [reply src/weathergen/train/loss_calculator.py]: Not sure if I tried -  let me test again.
Btw, I had the same issue [here](https://github.com/ecmwf/WeatherGenerator/pull/705/files#diff-7b5c433e5f3049bb43897cfc93a3d0cd1ffabeadde5a754a5f7a268bc9566428R139).

File: src/weathergen/train/loss_calculator.py (hunk 2 above)
Comment (Jubeku) [reply src/weathergen/train/loss_calculator.py]: Moved it to loss.py

File: src/weathergen/train/loss_calculator.py (hunk 1 above)
Comment (Jubeku) [reply src/weathergen/train/loss_calculator.py]: That works. Fixed it.

File: src/weathergen/train/loss_calculator.py (hunk 5 above)
Comment (Jubeku) [reply src/weathergen/train/loss_calculator.py]: I check now even before if `location_weight_type` (before named `stream_info_location_weight`) is None and return, this way I don't have to check afterwards. This also fixes `.to(device)` as it will never receive None. 
Checking early simplified the whole function, thanks.

File: src/weathergen/train/loss_calculator.py (hunk 4 above)
Comment (Jubeku) [reply src/weathergen/train/loss_calculator.py:114]: See comment in https://github.com/ecmwf/WeatherGenerator/pull/705#discussion_r2276111271

**Review by clessig — APPROVED** [review]
Ready to be merged.
